janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/step/JanusGraphMultiQueryStep.java

```diff
@@ -34,50 +34,72 @@
  * if initialised with all the starts than just one at a time, so this step allows it to
  * request the full set of starts from this step when initialising itself.
  */
-public final class JanusGraphMultiQueryStep extends AbstractStep<Vertex, Vertex> {
+public final class JanusGraphMultiQueryStep extends AbstractStep<Element, Element> {
 
-    private final Set<Traverser.Admin<Vertex>> cachedStarts = new HashSet<>();
-    private final String forStep;
-    private boolean cachedStartsAccessed = false;
+    private List<MultiQueriable> clientSteps = new ArrayList<>();
+    private final boolean limitBatchSize;
+    private boolean initialized;
 
-    public JanusGraphMultiQueryStep(Step<Vertex,?> originalStep) {
-        super(originalStep.getTraversal());
-        this.forStep = originalStep.getClass().getSimpleName();
+    public JanusGraphMultiQueryStep(Traversal.Admin traversal, boolean limitBatchSize) {
+        super(traversal);
+        this.limitBatchSize = limitBatchSize;
+        this.initialized = false;
     }
 
-    @Override
-    protected Admin<Vertex> processNextStart() throws NoSuchElementException {
-        Admin<Vertex> start = this.starts.next();
-        if (!cachedStarts.contains(start))
-        {
-            if (cachedStartsAccessed) {
-                cachedStarts.clear();
-                cachedStartsAccessed = false;
+    public void attachClient(MultiQueriable mq) {
+        clientSteps.add(mq);
+    }
+
+    private void initialize() {
+        assert !initialized;
+        initialized = true;
+
+        if (!limitBatchSize && !clientSteps.isEmpty()) { // eagerly cache all starts instead of batching
+            if (!starts.hasNext()) {
+                throw FastNoSuchElementException.instance();
             }
-            final List<Traverser.Admin<Vertex>> newStarters = new ArrayList<>();
-            starts.forEachRemaining(s -> {
-                newStarters.add(s);
-                cachedStarts.add(s);
+            final List<Traverser.Admin<Element>> elements = new ArrayList<>();
+            starts.forEachRemaining(e -> {
+                elements.add(e);
+                if (e.get() instanceof Vertex) {
+                    clientSteps.forEach(client -> client.registerFutureVertexForPrefetching((Vertex) e.get()));
```

Comment:
I will add the following documentation:

```java
    /**
     * All steps that use this step to fill their cache. For example, this could be the
     * next JanusGraphVertexStep. If the next step is a MultiQuery compatible parent
     * (such as union()), then all of its child traversals can use this cache. Thus,
     * there can be more than one client step.
     */
    private List<MultiQueriable> clientSteps = new ArrayList<>();
```

```java
    /**
     * Registers a vertex which will pass this step at some point in the future.
     * The vertex is typically known because a traverser at that vertex location
     * has passed a previous step earlier.
     * Using that information, a step can know in advance a set of vertices which
     * it will have to handle in the future.
     * @param futureVertex The vertex which will reach the step in the future.
     */
    void registerFutureVertexForPrefetching(Vertex futureVertex)
```